File: Clients/macOS/ValeniusShared/Sources/ValeniusShared/IpcMessages.swift
// Mirrors Valenius.Shared.IpcMessages (Shared/Valenius.Shared/IpcMessages.cs) and its
// Python port (Clients/Linux/shared/messages.py). No code generation across the repos —
// keep these three in lockstep by hand. JSON field names are PascalCase to match .NET's
// default serialization; CommandType serializes as its underlying Int, not a string.

import Foundation

public enum CommandType: Int, Codable, Sendable {
    case connect = 0
    case disconnect = 1
    case status = 2
    case uploadConfig = 3
    case getConfigInfo = 4
    case checkUpdate = 5
    case register = 6
    case syncStatus = 7
    case setAutoConnect = 8
    case deleteProfile = 9
    /// Sent by the app just before it exits so the daemon can notify the backend.
    case notifyOffline = 10
    /// Sent by the app to confirm MFA enrollment with a TOTP code (carried in `PipeCommand.mfaCode`).
    case mfaEnrollConfirm = 11
    /// Sent by the app's "Send logs" action — the daemon collects its (Valenius-only,
    /// redacted) diagnostic logs and uploads them to the backend.
    case sendLogs = 12
}

public struct PipeCommand: Codable, Sendable {
    public var command: CommandType
    public var configContent: String?
    public var profileName: String?
    public var autoConnectEnabled: Bool?
    /// TOTP code for `CommandType.mfaEnrollConfirm`.
    public var mfaCode: String?

    public init(
        command: CommandType,
        configContent: String? = nil,
        profileName: String? = nil,
        autoConnectEnabled: Bool? = nil,
        mfaCode: String? = nil
    ) {
        self.command = command
        self.configContent = configContent
        self.profileName = profileName
        self.autoConnectEnabled = autoConnectEnabled
        self.mfaCode = mfaCode
    }

    enum CodingKeys: String, CodingKey {
        case command = "Command"
        case configContent = "ConfigContent"
        case profileName = "ProfileName"
        case autoConnectEnabled = "AutoConnectEnabled"
        case mfaCode = "MfaCode"
    }
}

public struct PipeResponse: Codable, Sendable {
    public var success: Bool
    public var error: String?
    public var dataJson: String?

    public init(success: Bool, error: String? = nil, dataJson: String? = nil) {
        self.success = success
        self.error = error
        self.dataJson = dataJson
    }

    enum CodingKeys: String, CodingKey {
        case success = "Success"
        case error = "Error"
        case dataJson = "DataJson"
    }

    public static func ok(dataJson: String? = nil) -> PipeResponse {
        PipeResponse(success: true, dataJson: dataJson)
    }

    public static func ok<T: Encodable>(_ value: T) -> PipeResponse {
        let json = (try? String(data: JSONEncoder().encode(value), encoding: .utf8)) ?? nil
        return PipeResponse(success: true, dataJson: json)
    }

    public static func fail(_ message: String) -> PipeResponse {
        PipeResponse(success: false, error: message)
    }
}

/// One connected WireGuard tunnel. Multiple may be active simultaneously (the
/// primary/server VPN plus any number of foreign-customer profiles).
public struct ConnectedTunnelInfo: Codable, Sendable {
    public var name: String
    public var isVerified: Bool
    public var connectedUser: String?
    /// ISO-8601 string (matches .NET DateTime / Python isoformat on the wire).
    public var connectedSince: String?

    public init(name: String, isVerified: Bool = false, connectedUser: String? = nil, connectedSince: String? = nil) {
        self.name = name
        self.isVerified = isVerified
        self.connectedUser = connectedUser
        self.connectedSince = connectedSince
    }

    enum CodingKeys: String, CodingKey {
        case name = "Name"
        case isVerified = "IsVerified"
        case connectedUser = "ConnectedUser"
        case connectedSince = "ConnectedSince"
    }
}

public struct TunnelStatus: Codable, Sendable {
    public var isConnected: Bool = false
    /// True once the background probe confirmed end-to-end traffic through the tunnel.
    /// The UI replaces the green dot with a green checkmark when this is true.
    public var isVerified: Bool = false
    public var connectedUser: String?
    public var tunnelName: String?
    public var connectedSince: String?

    /// All currently-connected tunnels. The source of truth for per-profile row
    /// indicators in the menu bar app. The single-tunnel fields above mirror the
    /// primary/server profile entry here (or the first entry) for backward compat.
    public var connectedTunnels: [ConnectedTunnelInfo] = []
    public var registrationIsActive: Bool?
    public var hasStagedConfig: Bool = false
    public var updateAvailable: Bool = false
    public var availableProfiles: [String] = []
    public var deletableProfiles: [String] = []
    public var autoConnectEnabled: Bool = false

    // MARK: MFA session gating (Pro)

    /// True when this peer is MFA-gated and has no active session — the app prompts for re-auth.
    public var mfaRequired: Bool = false
    /// Deep link the app opens in the browser to authorize.
    public var mfaAuthorizeUrl: String?
    /// Absolute expiry of the active MFA session, for the countdown. Nil when none.
    public var mfaSessionExpiresAt: String?
    /// True when the TOTP enrollment window is open — the app shows the QR.
    public var mfaEnrollmentOpen: Bool = false
    /// otpauth:// URI for the enrollment QR.
    public var mfaEnrollmentUri: String?
    /// For a gated requester with a push-approver device: the number to display
    /// ("approve N on your phone"). Nil when there's no active push-approve challenge.
    public var mfaApproveNumber: Int?

    /// Daemon-derived (the app can't read the root-owned appsettings.json).
    public var backendUrl: String?
    public var backendReachable: Bool?

    public init() {}

    enum CodingKeys: String, CodingKey {
        case isConnected = "IsConnected"
        case isVerified = "IsVerified"
        case connectedUser = "ConnectedUser"
        case tunnelName = "TunnelName"
        case connectedSince = "ConnectedSince"
        case connectedTunnels = "ConnectedTunnels"
        case registrationIsActive = "RegistrationIsActive"
        case hasStagedConfig = "HasStagedConfig"
        case updateAvailable = "UpdateAvailable"
        case availableProfiles = "AvailableProfiles"
        case deletableProfiles = "DeletableProfiles"
        case autoConnectEnabled = "AutoConnectEnabled"
        case mfaRequired = "MfaRequired"
        case mfaAuthorizeUrl = "MfaAuthorizeUrl"
        case mfaSessionExpiresAt = "MfaSessionExpiresAt"
        case mfaEnrollmentOpen = "MfaEnrollmentOpen"
        case mfaEnrollmentUri = "MfaEnrollmentUri"
        case mfaApproveNumber = "MfaApproveNumber"
        case backendUrl = "BackendUrl"
        case backendReachable = "BackendReachable"
    }
}

public struct RegistrationResult: Codable, Sendable {
    public var isActive: Bool
    public var message: String?

    public init(isActive: Bool, message: String? = nil) {
        self.isActive = isActive
        self.message = message
    }

    enum CodingKeys: String, CodingKey {
        case isActive = "IsActive"
        case message = "Message"
    }
}

public struct ConfigInfo: Codable, Sendable {
    public var hasConfig: Bool = false
    public var tunnelName: String?
    public var fileName: String?
    public var profiles: [String] = []

    public init(hasConfig: Bool = false, tunnelName: String? = nil, fileName: String? = nil, profiles: [String] = []) {
        self.hasConfig = hasConfig
        self.tunnelName = tunnelName
        self.fileName = fileName
        self.profiles = profiles
    }

    enum CodingKeys: String, CodingKey {
        case hasConfig = "HasConfig"
        case tunnelName = "TunnelName"
        case fileName = "FileName"
        case profiles = "Profiles"
    }
}

public struct VersionCheckResult: Codable, Sendable {
    public var updateAvailable: Bool = false
    public var currentVersion: String?
    public var latestVersion: String?
    public var releaseNotes: String?

    public init(updateAvailable: Bool = false, currentVersion: String? = nil, latestVersion: String? = nil, releaseNotes: String? = nil) {
        self.updateAvailable = updateAvailable
        self.currentVersion = currentVersion
        self.latestVersion = latestVersion
        self.releaseNotes = releaseNotes
    }

    enum CodingKeys: String, CodingKey {
        case updateAvailable = "UpdateAvailable"
        case currentVersion = "CurrentVersion"
        case latestVersion = "LatestVersion"
        case releaseNotes = "ReleaseNotes"
    }
}

public enum ProfileNameHelper {
    public static func isValid(_ name: String?) -> Bool {
        guard let name, !name.isEmpty, name.count <= 50 else { return false }
        return name.allSatisfy { $0.isLetter || $0.isNumber || $0 == "_" || $0 == "-" }
    }

    public static func sanitize(_ filename: String) -> String {
        let stem = (filename as NSString).deletingPathExtension
        let safe = String(stem.map { $0.isLetter || $0.isNumber || $0 == "_" || $0 == "-" ? $0 : "_" })
        return String(safe.prefix(50))
    }
}

File: Shared/Valenius.Shared/IpcMessages.cs
namespace Valenius.Shared;

public enum CommandType
{
    Connect,
    Disconnect,
    Status,
    UploadConfig,
    GetConfigInfo,
    CheckUpdate,
    Register,
    SyncStatus,
    SetAutoConnect,
    DeleteProfile,
    /// <summary>Sent by the TrayApp just before it exits so the service can notify the backend.</summary>
    NotifyOffline,
    /// <summary>Sent by the TrayApp to confirm MFA enrollment with a TOTP code (carried in <see cref="PipeCommand.MfaCode"/>).</summary>
    MfaEnrollConfirm,
    /// <summary>Sent by the TrayApp "Send logs" action — the service collects its (Valenius-only,
    /// redacted) diagnostic logs and uploads them to the backend.</summary>
    SendLogs,
    /// <summary>Sent by the TrayApp first-run prompt to set the backend server address. The DNS name
    /// is carried in <see cref="PipeCommand.BackendDns"/>; the service prepends https:// and persists it.</summary>
    SetBackendUrl,
    /// <summary>Sent by the TrayApp's first-run "Use standalone" choice. No payload — the service just
    /// persists that the user opted out of backend setup, so the tray stops prompting and shows the
    /// normal popup (local profiles, no backend contacted) instead of the setup-required block.</summary>
    SetStandaloneMode
}

public class PipeCommand
{
    public CommandType Command { get; set; }
    public string? ConfigContent { get; set; }
    public string? ProfileName { get; set; }
    public bool? AutoConnectEnabled { get; set; }
    /// <summary>TOTP code for <see cref="CommandType.MfaEnrollConfirm"/>.</summary>
    public string? MfaCode { get; set; }
    /// <summary>DNS host for <see cref="CommandType.SetBackendUrl"/> (without scheme; the service prepends https://).</summary>
    public string? BackendDns { get; set; }
    /// <summary>For <see cref="CommandType.UploadConfig"/>: when true, the uploaded profile is also
    /// archived to the backend so every OS user on this machine can receive it, not just the
    /// uploading account. Defaults to (null/false) local-only, matching prior behavior.</summary>
    public bool? ShareWithAllUsers { get; set; }
}

public class PipeResponse
{
    public bool Success { get; set; }
    public string? Error { get; set; }
    public string? DataJson { get; set; }
    /// <summary>True when <see cref="Error"/> is a pre-connect local-LAN-conflict refusal
    /// (the client's own LAN overlaps a network reachable through the VPN, or the assigned VPN
    /// IP itself). The TrayApp shows a blocking modal for this instead of the usual balloon.</summary>
    public bool IsLanConflict { get; set; }
    /// <summary>For <see cref="CommandType.UploadConfig"/> with <see cref="PipeCommand.ShareWithAllUsers"/>:
    /// true when the local save succeeded (<see cref="Success"/> is still true) but archiving the
    /// profile to the backend for other OS users failed — <see cref="Error"/> carries a
    /// user-facing explanation. The profile stays usable, just not yet shared.</summary>
    public bool ShareFailed { get; set; }
}

/// <summary>One connected WireGuard tunnel. Multiple may be active simultaneously
/// (the primary/server VPN plus any number of foreign-customer profiles).</summary>
public class ConnectedTunnelInfo
{
    public string Name { get; set; } = "";
    public bool IsVerified { get; set; }
    /// <summary>
    /// True once the initial ~60s verification window elapsed with no success. Not terminal —
    /// verification keeps retrying at a slower cadence in the background, and a late success
    /// still clears this and sets <see cref="IsVerified"/>. The tray shows an amber
    /// "Confirmation failed" tag instead of the plain "Connected" one.
    /// </summary>
    public bool VerificationFailed { get; set; }
    /// <summary>
    /// The UDP port this tunnel switched to (typically 443) after the primary WireGuard® port
    /// produced no successful verification within the trigger window
    /// (docs/design/port-443-fallback.md), or 0 if it's still on the primary port. Sticky for the
    /// life of the connection — cleared only by a fresh connect, which always starts back on the
    /// primary port. The tray appends " · &lt;port&gt;" to whichever tag it would otherwise show,
    /// since it's useful for the user to know their traffic left on an alternate port.
    /// </summary>
    public int FallbackPortUsed { get; set; }
    public string? ConnectedUser { get; set; }
    public DateTime ConnectedSince { get; set; }
}

public class TunnelStatus
{
    public bool IsConnected { get; set; }
    /// <summary>
    /// True once the background probe confirmed end-to-end traffic through the tunnel.
    /// The UI replaces the green dot with a green checkmark when this is true.
    /// </summary>
    public bool IsVerified { get; set; }
    /// <summary>Mirrors <see cref="ConnectedTunnelInfo.VerificationFailed"/> for the primary/server tunnel.</summary>
    public bool VerificationFailed { get; set; }
    /// <summary>Mirrors <see cref="ConnectedTunnelInfo.FallbackPortUsed"/> for the primary/server tunnel.</summary>
    public int FallbackPortUsed { get; set; }
    public string? ConnectedUser { get; set; }
    public string? TunnelName { get; set; }
    public DateTime? ConnectedSince { get; set; }

    /// <summary>
    /// All currently-connected tunnels. The source of truth for per-profile row
    /// indicators in the tray. The single-tunnel fields above mirror the primary/server
    /// profile entry here (or the first entry) for backward compatibility with the status
    /// bar, MFA flow, and the old no-profile disconnect path.
    /// </summary>
    public ConnectedTunnelInfo[] ConnectedTunnels { get; set; } = [];
    public bool? RegistrationIsActive { get; set; }
    public bool HasStagedConfig { get; set; }
    public bool UpdateAvailable { get; set; }

    /// <summary>
    /// True when no backend server address is configured yet (fresh install without an installer-provided
    /// URL) — the tray prompts the user for the server DNS. Intentionally phrased so that a missing field
    /// from an older service deserializes to <c>false</c> (configured) and never triggers a spurious prompt.
    /// </summary>
    public bool BackendUnconfigured { get; set; }
    /// <summary>
    /// True once the user explicitly chose "Use standalone" in the first-run setup dialog while
    /// <see cref="BackendUnconfigured"/> is true. The tray shows the normal popup (local profiles,
    /// connect/upload/delete all work with no backend contacted) instead of the setup-required
    /// block, plus a "Connect to a server" action to opt in later. Default <c>false</c> preserves
    /// today's behavior — same reasoning as <see cref="BackendUnconfigured"/>'s doc comment.
    /// </summary>
    public bool StandaloneMode { get; set; }
    public string[] AvailableProfiles  { get; set; } = [];
    public string[] DeletableProfiles  { get; set; } = [];
    public bool     AutoConnectEnabled { get; set; }

    // ── MFA session gating (Pro) ──────────────────────────────────────────────
    /// <summary>True when this peer is MFA-gated and has no active session — the tray prompts for re-auth.</summary>
    public bool      MfaRequired         { get; set; }
    /// <summary>Deep link the tray opens in the browser to authorize.</summary>
    public string?   MfaAuthorizeUrl     { get; set; }
    /// <summary>Absolute expiry of the active MFA session, for the countdown. Null when none.</summary>
    public DateTime? MfaSessionExpiresAt { get; set; }
    /// <summary>True when the TOTP enrollment window is open — the tray shows the QR.</summary>
    public bool      MfaEnrollmentOpen   { get; set; }
    /// <summary>otpauth:// URI for the enrollment QR.</summary>
    public string?   MfaEnrollmentUri    { get; set; }
    /// <summary>For a gated requester with a push-approver device: the number to display
    /// ("approve N on your phone"). Null when there's no active push-approve challenge.</summary>
    public int?      MfaApproveNumber    { get; set; }
}

public class RegistrationResult
{
    public bool IsActive { get; set; }
    public string? Message { get; set; }
}

public class ConfigInfo
{
    public bool HasConfig { get; set; }
    public string? TunnelName { get; set; }
    public string? FileName { get; set; }
    public string[] Profiles { get; set; } = [];
}

public static class ProfileNameHelper
{
    public static bool IsValid(string? name) =>
        !string.IsNullOrEmpty(name) &&
        name.Length <= 50 &&
        name.All(c => char.IsLetterOrDigit(c) || c == '_' || c == '-');

    public static string Sanitize(string filename)
    {
        var stem = Path.GetFileNameWithoutExtension(filename);
        var safe = new string(stem.Select(c => char.IsLetterOrDigit(c) || c == '_' || c == '-' ? c : '_').ToArray());
        return safe.Length > 50 ? safe[..50] : safe;
    }
}

public class VersionCheckResult
{
    public bool UpdateAvailable { get; set; }
    public string? CurrentVersion { get; set; }
    public string? LatestVersion { get; set; }
    public string? ReleaseNotes { get; set; }
}
